api/plugin/vproto/api.proto

```diff
@@ -0,0 +1,317 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+syntax = "proto3";
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+import "google/protobuf/empty.proto";
+import "k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto";
+
+package vproto;
+
+option go_package = "github.com/containerd/nri/api/plugin/vproto;vproto";
+
+option (gogoproto.goproto_unkeyed_all) = false;
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+option (gogoproto.goproto_getters_all) = true;
+
+
+//
+// Plugin is the ttRPC service than NRI plugins must implement.
+//
+service Plugin {
+  // Configure the plugin.
+  //
+  // Static plugins, IOW plugins that are started by NRI, get passed any
+  // plugin-specific raw configuration data found in the NRI configuration.
+  // Dynamic plugins, IOW plugins started by means external to NRI that
+  // register themselves by connecting to the well-known NRI socket, get
+  // empty configuration data.
+  //
+  // A plugin responds to a configuration request with the set of events it
+  // wants to subscribe to.
+  //
+  rpc Configure(ConfigureRequest) returns (ConfigureResponse);
+
+  // Synchronize the plugin with the state of the runtime.
+  //
+  // Upon startup NRI synchronizes the state of the plugin with the runtime
+  // by sending a list of all pods and containers. The plugin responds
+  // with any changes it wishes to make to the containers.
+  //
+  rpc Synchronize(SynchronizeRequest) returns (SynchronizeResponse);
+
+  // Shutdown lets the plugin know the runtime is going down.
+  //
+  // Once a plugin has responded to this request it should shut itself down.
```

Comment:
It seems `Shutdown` is called when the runtime thinks a plugin should not continue due to an error, too. That is, not only when the runtime is going down as documented.

I think there is a difference in these cases, especially in case of dynamic plugins whose lifecycle is not managed by the runtime. A process that acts as a dynamic plugin (and possibly does something else as well, and is possibly managed by systemd or Kubernetes) should be able to tell if the runtime wants it to go away due to an error and never call back, or if the admin is just doing `systemctl restart containerd` and the plugin process could just wait a while and then start trying to reconnect to `nri.sock`.

Reply:
> It seems `Shutdown` is called when the runtime thinks a plugin should not continue due to an error, too. That is, not only when the runtime is going down as documented.

Yes, or to be more precise it is used after errors considered fatal (for the continued operation of the plugin) to let the plugin know that the runtime/NRI is closing the connection.

> I think there is a difference in these cases, especially in case of dynamic plugins whose lifecycle is not managed by the runtime. A process that acts as a dynamic plugin (and possibly does something else as well, and is possibly managed by systemd or Kubernetes) should be able to tell if the runtime wants it to go away due to an error and never call back, or if the admin is just doing `systemctl restart containerd` and the plugin process could just wait a while and then start trying to reconnect to `nri.sock`.

Yes, true. This is far from being properly thought through at all. The basic idea was just roughly that if the error leading to the NRI's decision to kick out the plugin was considered fatal by the plugin as well, it will probably exit too, then eventually (maybe after manual intervention by an admin to fix the source of error), get restarted, reinitialize itself, reconnect and get resynchronized. And if the plugin considers the error to be transient, it can try the same reconnect-resynchronize sequence without exiting first. And if the  shutting down is due to the runtime being shut down, the plugin won't know if/when the runtime is being started up again, so it will just keep trying to reconnect. That's why there are no details in the message about why the connection is being closed: it seems to me that it won't really change how the plugin reacts as the reaction is more a function of the state of the plugin itself. But it definitely wouldn't hurt the plugin to know so we can as well add it.